org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeLensHandler.java

```diff
@@ -11,150 +11,51 @@
 package org.eclipse.jdt.ls.core.internal.handlers;
 
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
 import org.eclipse.core.runtime.IProgressMonitor;
-import org.eclipse.jdt.core.Flags;
+import org.eclipse.core.runtime.Platform;
 import org.eclipse.jdt.core.ICompilationUnit;
 import org.eclipse.jdt.core.IJavaElement;
-import org.eclipse.jdt.core.IJavaProject;
-import org.eclipse.jdt.core.ISourceRange;
-import org.eclipse.jdt.core.ISourceReference;
-import org.eclipse.jdt.core.IType;
-import org.eclipse.jdt.core.JavaCore;
 import org.eclipse.jdt.core.JavaModelException;
-import org.eclipse.jdt.core.search.IJavaSearchConstants;
-import org.eclipse.jdt.core.search.IJavaSearchScope;
-import org.eclipse.jdt.core.search.SearchEngine;
-import org.eclipse.jdt.core.search.SearchMatch;
-import org.eclipse.jdt.core.search.SearchParticipant;
-import org.eclipse.jdt.core.search.SearchPattern;
-import org.eclipse.jdt.core.search.SearchRequestor;
 import org.eclipse.jdt.ls.core.internal.JDTUtils;
 import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
-import org.eclipse.jdt.ls.core.internal.ResourceUtils;
 import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
 import org.eclipse.lsp4j.CodeLens;
-import org.eclipse.lsp4j.Command;
-import org.eclipse.lsp4j.Location;
-import org.eclipse.lsp4j.Range;
 
 public class CodeLensHandler {
 
-	private static final String JAVA_SHOW_REFERENCES_COMMAND = "java.show.references";
-	private static final String JAVA_SHOW_IMPLEMENTATIONS_COMMAND = "java.show.implementations";
-	private static final String IMPLEMENTATION_TYPE = "implementations";
-	private static final String REFERENCES_TYPE = "references";
-
 	private final PreferenceManager preferenceManager;
+	private List<CodeLensProvider> codeLensProviders;
 
 	public CodeLensHandler(PreferenceManager preferenceManager) {
 		this.preferenceManager = preferenceManager;
+		codeLensProviders = new ArrayList<>();
+		codeLensProviders.add(new ReferencesCodeLensProvider(preferenceManager));
```

Comment:
Why not add `ReferencesCodeLensProvider` and `ImplementationCodeLensProvider` as extensions rather than a coded way.

Reply:
I thought about it too, in that case, we'll need a priority attribute to ensure consistent codelens order

Reply:
I planned to add junit codelens provider as an extension and "Implementation" , "Reference"  are more like language server embedded providers.  About the order, if provider extensions don't know each other's existence, how could a codelens provider to decide its priority?

Reply:
See similar concept in https://github.com/eclipse/eclipse.jdt.ls/blob/3d5e4d1aeae9c29ed5a7ca49ae28f1903c51c44d/org.eclipse.jdt.ls.core/plugin.xml#L59-L74

The jdt.ls ecosystem shouldn't be so big that you couldn't check what other providers use.